src/overlays/actors/ovl_En_Bal/z_en_bal.c

```diff
@@ -64,108 +149,1026 @@ static DamageTable D_80A63CE0 = {
+            break;
+
+        case TINGLE_MAP_WOODFALL:
+            SET_WEEKEVENTREG(WEEKEVENTREG_BOUGHT_TINGLE_MAP_WOODFALL);
+            break;
+
+        case TINGLE_MAP_SNOWHEAD:
+            SET_WEEKEVENTREG(WEEKEVENTREG_BOUGHT_TINGLE_MAP_SNOWHEAD);
+            break;
+
+        case TINGLE_MAP_ROMANI_RANCH:
+            SET_WEEKEVENTREG(WEEKEVENTREG_BOUGHT_TINGLE_MAP_ROMANI_RANCH);
+            break;
+
+        case TINGLE_MAP_GREAT_BAY:
+            SET_WEEKEVENTREG(WEEKEVENTREG_BOUGHT_TINGLE_MAP_GREAT_BAY);
+            break;
+
+        case TINGLE_MAP_STONE_TOWER:
+            SET_WEEKEVENTREG(WEEKEVENTREG_BOUGHT_TINGLE_MAP_STONE_TOWER);
+            break;
+    }
+}
+
+void EnBal_TryPurchaseMap(EnBal* this, PlayState* play) {
+    Player* player = GET_PLAYER(play);
+    s32 price;
+
+    if (Message_ShouldAdvance(play)) {
+        if (play->msgCtx.choiceIndex != EN_BAL_MAPCHOICE_CANCEL) {
+            // Get price depending on which map player wants to buy
+            if (play->msgCtx.choiceIndex == EN_BAL_MAPCHOICE_PROXIMAL) {
+                price = play->msgCtx.unk1206C;
+            } else {
+                price = play->msgCtx.unk12070;
+            }
+
+            if (gSaveContext.save.saveInfo.playerData.rupees < price) {
+                // Can't buy map because player doesn't have the money
+                Audio_PlaySfx(NA_SE_SY_ERROR);
+                Actor_ChangeAnimationByInfo(&this->skelAnime, sAnimationInfo, TINGLE_ANIM_TALK_IDLE);
+                Message_StartTextbox(play, 0x1D0A, &this->picto.actor);
+                this->prevTextId = 0x1D0A;
+            } else if (EnBal_CheckIfMapUnlocked(this, play)) {
+                // Can't buy map because player already has it
+                Audio_PlaySfx(NA_SE_SY_ERROR);
+                Actor_ChangeAnimationByInfo(&this->skelAnime, sAnimationInfo, TINGLE_ANIM_TALK_IDLE);
+                Message_StartTextbox(play, 0x1D09, &this->picto.actor);
+                this->prevTextId = 0x1D09;
+            } else {
+                // Proceed with map purchase
+                Audio_PlaySfx_MessageDecide();
+                Rupees_ChangeBy(-price);
+                Actor_ChangeAnimationByInfo(&this->skelAnime, sAnimationInfo, TINGLE_ANIM_MAGIC_SHORT);
+                this->forceEyesShut = true;
+                Message_StartTextbox(play, 0x1D0B, &this->picto.actor);
+                this->prevTextId = 0x1D0B;
+                EnBal_UnlockSelectedAreaMap(this);
+                player->stateFlags1 |= PLAYER_STATE1_20;
+                EnBal_SetupOfferGetItem(this);
+            }
+        } else {
+            // Cancel
+            Audio_PlaySfx_MessageCancel();
+            Actor_ChangeAnimationByInfo(&this->skelAnime, sAnimationInfo, TINGLE_ANIM_TALK_IDLE);
+            Message_StartTextbox(play, 0x1D06, &this->picto.actor);
+            this->prevTextId = 0x1D06;
+        }
+    }
+}
+
+void EnBal_ProceedWithDialogue(EnBal* this, PlayState* play) {
```

Comment:
Perhaps it is just me, but I've preferred `HandleConversation` for functions like this.